File: 2023/Kotlin/app/src/main/kotlin/org/example/nathandow/aoc/Day1.kt
package org.example.nathandow.aoc

import java.io.File

class Day1 {
    /*
     * Note: This is valid for both part 1 & 2.
     */
    fun run(inputFile: String) : Int {
        val inputStream = File(inputFile).inputStream()
        return inputStream
            .bufferedReader()
            .readLines()
            .fold(0) { sum, line -> sum + firstLast(line) }
    }

    private fun firstLast(str: String) : Int {
        var s = 0
        var e = 0

        for (i in str.indices) {
            if (s == 0) { s = getDigit(str, i) }
            if (e == 0) { e = getDigit(str, str.length - (i + 1)) }

            if (s != 0 && e != 0) {
                return "$s$e".toInt()
            }
        }

        return 0
    }

    private fun getDigit(str: String, i: Int) : Int {
        return if (str[i].isDigit()) {
            str[i].digitToInt()
        } else {
            getAlphaDigit(str, i)
        }
    }

    private fun getAlphaDigit(str: String, i: Int) : Int {
        val englishToDigit = mapOf(
            "one" to 1,
            "two" to 2,
            "three" to 3,
            "four" to 4,
            "five" to 5,
            "six" to 6,
            "seven" to 7,
            "eight" to 8,
            "nine" to 9
        )

        for (k in englishToDigit.keys) {
            if (str.substring(i).startsWith(k)) {
                return englishToDigit[k] ?: 0
            }
        }

        return 0
    }
}

File: 2023/Swift/AoC/AoC/Day1.swift
//
//  Day1.swift
//  AoC
//
//  Created by Nathaniel Downey on 7/11/24.
//

import Foundation

class Day1 {
    private let englishToDigit = ["one": 1, "two": 2, "three": 3, "four": 4, "five": 5, "six": 6, "seven": 7, "eight": 8, "nine": 9 ]
    
    func run(data: String) -> Int {
        var out = 0
        let lines = data.split(separator: "\n")
        for subLine in lines {
            let line = String(subLine)
            let firstLast = combineFirstLastDigits(str: line)
            out += firstLast
        }
        
        return out
    }
    
    private func combineFirstLastDigits(str: String) -> Int {
        var first = 0
        var last = 0
        
        for index in str.indices {
            if first != 0 { break }
            first = asDigit(str: str, index: index)
        }
        
        for index in str.indices.reversed() {
            if last != 0 { break }
            last = asDigit(str: str, index: index)
        }
        
        return (first * 10) + last
    }
    
    private func asDigit(str: String, index: String.Index) -> Int {
        if str[index].isNumber { return str[index].wholeNumberValue ?? 0 }
        return englishAsDigit(str: str, index: index)
    }
    
    private func englishAsDigit(str: String, index: String.Index) -> Int {
        for key in englishToDigit.keys {
            if str[index...].starts(with: key) {
                if let value = englishToDigit[key] {
                    return value
                }
            }
        }
        
        return 0
    }
}
